tests/cpgrid/adapt_cpgrid_test.cpp

```diff
@@ -0,0 +1,326 @@
+//
+// $Date$
+//
+// $Revision$
+//
+//===========================================================================
+/*
+  Copyright 2024 Equinor ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include "config.h"
+
+#define BOOST_TEST_MODULE AdaptTests
+#include <boost/test/unit_test.hpp>
+#include <boost/version.hpp>
+#if BOOST_VERSION / 100000 == 1 && BOOST_VERSION / 100 % 1000 < 71
+#include <boost/test/floating_point_comparison.hpp>
+#else
+#include <boost/test/tools/floating_point_comparison.hpp>
+#endif
+#include <opm/grid/CpGrid.hpp>
+#include <opm/grid/cpgrid/CpGridData.hpp>
+#include <opm/grid/cpgrid/DefaultGeometryPolicy.hpp>
+#include <opm/grid/cpgrid/Entity.hpp>
+#include <opm/grid/cpgrid/EntityRep.hpp>
+#include <opm/grid/cpgrid/Geometry.hpp>
+#include <opm/grid/LookUpData.hh>
+
+#include <dune/grid/common/mcmgmapper.hh>
+
+#include <sstream>
+#include <iostream>
+
+struct Fixture
+{
+    Fixture()
+    {
+        int m_argc = boost::unit_test::framework::master_test_suite().argc;
+        char** m_argv = boost::unit_test::framework::master_test_suite().argv;
+        Dune::MPIHelper::instance(m_argc, m_argv);
+        Opm::OpmLog::setupSimpleDefaultLogging();
+    }
+
+    static int rank()
+    {
+        int m_argc = boost::unit_test::framework::master_test_suite().argc;
+        char** m_argv = boost::unit_test::framework::master_test_suite().argv;
+        return Dune::MPIHelper::instance(m_argc, m_argv).rank();
+    }
+};
+
+BOOST_GLOBAL_FIXTURE(Fixture);
+
+void markAndAdapt_check(Dune::CpGrid& coarse_grid,
+                        const std::array<int,3>& cells_per_dim,
+                        const std::vector<int>& markedCells,
+                        Dune::CpGrid& other_grid,
+                        bool isBlockShape)
+{
+    int refinedLevel = coarse_grid.chooseData().size(); // size before calling adapt
+
+    for (const auto& elemIdx : markedCells)
+    {
+        const auto& elem =  Dune::cpgrid::Entity<0>(*(coarse_grid.chooseData()[0]), elemIdx, true);
+        coarse_grid.mark(1, elem);
+        coarse_grid.getMark(elem);
+        BOOST_CHECK( coarse_grid.getMark(elem) == 1);
+        BOOST_CHECK( elem.mightVanish() == true);
+    }
+    bool preAdapt = coarse_grid.preAdapt();
+    const auto& data = coarse_grid.chooseData();
+    if(preAdapt) {
```

Comment:
I see your point in this comment and the one above. I guess it would be better when preAdapt() takes markedCells (a vector with the selected indices to be refined) as a function argument (instead of manually -for-loop- marking each cell). 

Thanks for your first feedback!